### DISCUSSION TITLE
How to skip a test if a particular error is thrown?

### DISCUSSION THREAD

**Discussion by gajus** [vitest-dev/vitest#4318] [Q&A] [answered]
This is a bit of unusual use case, but basically I want to add runtime guards deep into codebase that would allow to conditionally skip the test.

The context is that we have a pretty intricate test setup. Some of these tests call APIs that should be only allowed to be called with a certain env variable is set. When that variable is not set, those tests should be just skipped.

In short, is there a way to analyze the error thrown inside of a test and conditionally skip the test?

**Comment by sheremet-va** [comment, marked as answer]
You can skip a test with `skip` function from the context:

```ts
test('some test', ({ skip }) => {
  skip()
})
```

It throws a `PendingError` which we don't expose anywhere. You can technically store this function in `beforeEach` and call it whenever you need to from outside the test:


```ts
beforeEach({ skip }) => {
  globalThis.skipCurrentTest = skip
})

// skipCurrentTest() anywhere in the codebase
```

**Reply by gajus** [reply]
`globalThis.skipCurrentTest` this would work

**Reply by gajus** [reply]
Is there any way to silence these errors?

```
stderr | testV2/integration/brandCustomization/invoice.test.ts > brand customization on invoices > successfully adds logo to invoice
PendingError: test is skipped; abort execution
    at context.skip (file:///Users/gajus/Developer/contra/gaia/node_modules/.pnpm/@vitest+runner@0.34.6/node_modules/@vitest/runner/dist/index.js:77:11)
    at globalThis.skipStripeTest (/Users/gajus/Developer/contra/gaia/apps/contra-api/test/vitest.setup.ts:68:5)
    at createNewUserSession (/Users/gajus/Developer/contra/gaia/packages/temporary-common/src/factories/helpers/createSessionFactory.ts:279:16)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at /Users/gajus/Developer/contra/gaia/packages/temporary-common/src/factories/helpers/createSessionFactory.ts:457:12
    at /Users/gajus/Developer/contra/gaia/apps/contra-api/testV2/integration/brandCustomization/invoice.test.ts:130:28
    at /Users/gajus/Developer/contra/gaia/apps/contra-api/testV2/helpers/createRunWithTestEnvironment.ts:193:15
    at /Users/gajus/Developer/contra/gaia/node_modules/.pnpm/@temporalio+worker@1.8.6/node_modules/@temporalio/worker/src/worker.ts:1676:16
    at async Promise.allSettled (index 0)
    at Worker.runUntil (/Users/gajus/Developer/contra/gaia/node_modules/.pnpm/@temporalio+worker@1.8.6/node_modules/@temporalio/worker/src/worker.ts:1683:44) {
  code: 'VITEST_PENDING',
  taskId: '-1209289037_0_1'
}
```

**Reply by sheremet-va** [reply]
They are silent by default. You are probably calling it somewhere where it can't be caught, and log it yourself (from what I see in your example).

**Reply by sheremet-va** [reply]
But there is also an option that can silence any log: https://vitest.dev/config/#onconsolelog

**Reply by gajus** [reply]
Thank you

**Reply by gajus** [reply]
They do get skipped for what it is worth.
